vault/expiration.go

```diff
@@ -245,125 +279,78 @@ func (m *ExpirationManager) Restore() error {
-	wg := &sync.WaitGroup{}
-
-	// Create 64 workers to distribute work to
-	for i := 0; i < consts.ExpirationRestoreWorkerCount; i++ {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-
-			for {
-				select {
-				case leaseID, ok := <-broker:
-					// broker has been closed, we are done
-					if !ok {
-						return
-					}
-
-					le, err := m.loadEntry(leaseID)
-					if err != nil {
-						errs <- err
-						continue
-					}
-
-					// Write results out to the result channel
-					result <- le
-
-				// quit early
-				case <-quit:
-					return
-				}
-			}
-		}()
-	}
-
-	// Distribute the collected keys to the workers in a go routine
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for i, leaseID := range existing {
-			if i%500 == 0 {
-				m.logger.Trace("expiration: leases loading", "progress", i)
-			}
-
-			select {
-			case <-quit:
-				return
-
-			default:
-				broker <- leaseID
-			}
-		}
-
-		// Close the broker, causing worker routines to exit
-		close(broker)
-	}()
-
 	// Restore each key by pulling from the result chan
-	for i := 0; i < len(existing); i++ {
-		select {
-		case err := <-errs:
-			// Close all go routines
-			close(quit)
+	for i, leaseID := range existing {
+		if i > 0 && i%500 == 0 {
+			m.logger.Trace("expiration: leases loading", "progress", i)
+		}
 
+		if err := m.loadAndRestoreLease(leaseID); err != nil {
 			return err
+		}
 
-		case le := <-result:
+		// This exists for test only to add latency to load calls
+		if loadDelay > 0 {
+			time.Sleep(loadDelay)
+		}
+	}
 
-			// If there is no entry, nothing to restore
-			if le == nil {
-				continue
-			}
+	// Clear our the restored entries and turn off restore mode
+	m.restoreLock.Lock()
+	m.restoreLoaded = nil
+	atomic.StoreInt64(&m.restoreMode, 0)
+	m.restoreLock.Unlock()
 
-			// If there is no expiry time, don't do anything
-			if le.ExpireTime.IsZero() {
-				continue
-			}
+	m.logger.Info("expiration: lease restore complete")
+	return nil
+}
 
-			// Determine the remaining time to expiration
-			expires := le.ExpireTime.Sub(time.Now())
-			if expires <= 0 {
-				expires = minRevokeDelay
-			}
+func (m *ExpirationManager) loadAndRestoreLease(leaseID string) error {
+	m.restoreLock.Lock()
+	defer m.restoreLock.Unlock()
 
-			// Setup revocation timer
-			m.pending[le.LeaseID] = time.AfterFunc(expires, func() {
-				m.expireID(le.LeaseID)
-			})
-		}
+	// If we loaded a lease elsewhere in the expiration manager while in
+	// restore, skip the entry since it may have been updated since we
+	// loaded it
+	if _, ok := m.restoreLoaded[leaseID]; ok {
+		return nil
 	}
 
-	// Let all go routines finish
-	wg.Wait()
-
-	if len(m.pending) > 0 {
-		if m.logger.IsInfo() {
-			m.logger.Info("expire: leases restored", "restored_lease_count", len(m.pending))
-		}
+	_, err := m.loadEntryInternal(leaseID, true)
+	if err != nil {
+		return err
 	}
 
 	return nil
 }
 
+// restoreLease takes a lease entry that has not been added to the expiration
+// manager and adds it back in
```

Comment:
Please add a note here that the pending lock needs to be held before calling this function.

Reply:
Oops, restore lock.